[Mod Content] parse mod data and save to all_mods.json

esphas opened the pull request:
- moved the data parse logic from `pull-data.mjs` to `parser/parser.mjs` and added support for mod data.
- added `modlist` entry to all.json to record the list of mods supported by that version.
- mod data is stored in all_mods.json.

check [esphas/cdda-data:main](https://github.com/esphas/cdda-data/tree/main) for examples of generated result.

nornagon commented:
Thanks for this, excited to take a look! However, this is hard for me to review given the file move; I can't really see what the changes were. I'm not opposed to reorganization, but can we have it not be in the same PR that makes significant changes to the code?

esphas commented:
> Thanks for this, excited to take a look! However, this is hard for me to review given the file move; I can't really see what the changes were. I'm not opposed to reorganization, but can we have it not be in the same PR that makes significant changes to the code?

@nornagon I've made some cleanup and test it with cdda-experimental-2025-08-25-1435, hope this helps.

nornagon commented on pull-data.mjs:
```diff
@@ -210,49 +215,81 @@ export default async function run({ github, context, dryRun = false }) {
-        obj.__filename = filename + `#L${start}-L${end}`;
-        data.push(obj);
+    for (const f of globFn("data/json/**/*.json")) {
+        const filename = f.name.replaceAll("\\", "/");
```
might as well pull this into `glob` too

nornagon commented on pull-data.mjs:
```diff
@@ -210,49 +215,81 @@ export default async function run({ github, context, dryRun = false }) {
+    const dataMods = {};
+    for (const i of globFn("data/mods/*/modinfo.json")) {
+      const modname = i.name.replaceAll("\\", "/").split("/")[2];
+      if (modname === "dda") continue;
```
i'd prefer to ignore the mod directory name completely and only read the mod id; can we change this to be a gate on `modId` instead?

and, come to think of it, i think we should also include the MOD_INFO for the dda mod in the bundle :) we can exclude it at display time.

nornagon commented on pull-data.mjs:
```diff
@@ -210,49 +215,81 @@ export default async function run({ github, context, dryRun = false }) {
       data,
-    };
-    const allJson = JSON.stringify(all);
+      modlist: Object.fromEntries(Object.entries(dataMods).map(([name, mod]) => [name, mod.info])),
```
mega nit:
```suggestion
      mods: Object.fromEntries(Object.entries(dataMods).map(([name, mod]) => [name, mod.info])),
```

it's not a list!

nornagon reviewed (commented):
looks great, thanks for reorganizing! just a couple of comments.

esphas commented:
@nornagon changes made:
- moved `.replaceAll("\\", "/")` into `globFn`
- removed the `"dda"` gate, we'll be handling that in the guide
- renamed `modlist` to `mods`
- added logic to copy `zh_xx_pinyin.json` to `latest/`, fixing the use of outdated pinyin files for the latest version

nornagon commented on pull-data.mjs:
```diff
@@ -354,11 +353,18 @@ export default async function run({ github, context, dryRun = false }) {
         `data/${latestBuild.build_number}/lang/${lang}.json`,
         `data/latest/lang/${lang}.json`,
       );
+      if (lang.startsWith("zh_")) {
+        copyBlob(
+          `data/${latestBuild.build_number}/lang/${lang}_pinyin.json`,
+          `data/latest/lang/${lang}_pinyin.json`,
+        );
+      }
```
thanks for this!

nornagon commented:
@esphas I'm going to merge this now, but it would also to be good to backfill the stable releases as they won't have mod info as-is. In the past I've done backfills manually when I've changed the data, but the repo is too big to clone now, so we'll have to do it in code 😅
